### PR TITLE
New LevMarq implementation

### PR DESCRIPTION
**Pull request by savuor**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/942
**Merge with contrib**: https://github.com/opencv/opencv_contrib/pull/3127

### TODOs:
- [x] new interface: constructors, `run(params)` for linear, various factories, return report after optimization
- [x] (optional) implement geodesic acceleration
- [x] play with the params on existing code to find better combinations
- [x] add perf reports
- [x] remove old impl
- [x] remove all `DEBUG` code and fulfill all `TODO` statements

### What?
New [Levenberg-Marquadt algorithm](https://en.wikipedia.org/wiki/Levenberg%E2%80%93Marquardt_algorithm) implementation replaces the old one.

### Why?
Because the old impl isn't as good as we want:
* No sparse matrix support
* No SO(3), SE(3) or any other manifold type supported for optimized params
* No flexibility: fixed up/down coefficients and some deltas, a few options for a termination criteria

This stops it from using in some complex 3d-related problems such as pose graph optimization.

An algorithm termination deserves its own list of _features_:
* Iteration calculation isn't fair: only successful ones are counted, a real amount of iterations elapsed can be several times bigger than reported
* The optimization can stop and be considered successful even regardless of successful iterations elapsed count, cost function value drop or NaNs in a gradient
* Step norm threshold is tied up to energy threshold

This is not OK even for a basic LevMarq solver.

### How?
A new code fixes everything:
* A base class does not depend on a type, layout or a group structure of a param vector or an objective function jacobian. A child class should provide a storage for that data and implement all virtual member functions that process it. This lets a user to use sparse matrices, exponential increments or fixed variables.
* The algorithm is highly tunable:
  - Lambda initial/up/down values can be changed by a user, diagonal clamping or upFactor doubling can be turned on/off
  - A termination criteria can be composed of the following ones, each threshold is independent and tunable:
    + relative energy delta < threshold
    + gradient max value < threshold
    + step norm (L2 or Inf) < threshold
    + cost function < threshold
    + iterations count < maxIterations
* Jacobian scaling, step quality metric and geodesic acceleration supported, they can improve the algorithm's speed/stability sometimes
* A dense linear implementation converges at least as good as the old code does

This code is based on a current pose graph optimization routine. As a result, the pose graph has been rewritten too, now it uses the same LevMarq impl as other OpenCV functions.

### Any numbers?

**TLDR:** new implementation converges more often in less iterations to approximately the same cost function values.

A convergence comparison across various use cases:

<details>
<summary>Convergence by function</summary>

|function|old good|old total|new good|new total|
|---|:-:|:-:|:-:|:-:|
| calibrateCameraInternal | 15 | 15 | 15 | 15 |
| estimateAffine2D | 11493 | 12111 | 12172 | 12172 |
| solvePnPRefine | 5 | 5 | 5 | 5 |
| stereoCalibrateImpl | 2 | 3 | 2 | 3 |
| estimateAffinePartial2D | 4300 | 6096 | 6858 | 6858 |
| findExtrinsicCameraParams2 | 6018 | 6341 | 6238 | 6329 |
| BundleAdjusterBase::estimate | 783 | 783 | 709 | 709 |
| findHomography | 25044 | 30957 | 31351 | 31719 |

</details>
<details>
<summary>Final energy by function</summary>

|function|old min|new min|old max|new max |old avg|new avg|old med|new med|
|---|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
| calibrateCameraInternal | 0.008321 | 0.008317 | 558.9 | 558.9 | 73.67 | 73.67 | 4.726 | 4.726 |
| estimateAffine2D | 0 | 0 | 3.267e+07 | 3.267e+07 | 7.014e+04 | 5.368e+05 | 0.009058 | 0.009339 |
| solvePnPRefine | 1.05e-26 | 2.985e-12 | 0.004971 | 0.004971 | 0.0009941 | 0.0009941 | 3.506e-17 | 2.179e-11 |
| stereoCalibrateImpl | 6.788 | 6.788 | 318.8 | 315.1 | 110.8 | 109.5 | 6.788 | 6.788 |
| estimateAffinePartial2D | 0 | 0 | 1209 | 1209 | 115.8 | 108.4 | 6.931e-09 | 3.247e-09 |
| findExtrinsicCameraParams2 | 1.883e-29 | 2.958e-31 | 2.729e+42 | 2.729e+42 | 4.304e+38 | 4.313e+38 | 2.898e-10 | 4.199e-10 |
| BundleAdjusterBase::estimate | 26 | 26 | 4.598e+04 | 4.598e+04 | 1869 | 2274 | 266.2 | 764.6 |
| findHomography | 4.645e-27 | 0 | 2304 | 2304 | 34 | 30.6 | 0.003876 | 3.122e-05 |

</details>
<details>
<summary>Iterations elapsed till convergence by function</summary>

|function|old min|new min|old max|new max|old avg|new avg|old med|new med|
|---|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
| calibrateCameraInternal | 5 | 4 | 157 | 54 | 60.07 | 24.73 | 70 | 26 |
| estimateAffine2D | 1 | 1 | 21 | 3 | 10.39 | 1.838 | 6 | 2 |
| solvePnPRefine | 3 | 2 | 7 | 4 | 5 | 3.6 | 5 | 4 |
| stereoCalibrateImpl | 45 | 1 | 71 | 1 | 58 | 1 | 58 | 1 |
| estimateAffinePartial2D | 1 | 1 | 21 | 6 | 4.561 | 1.655 | 4 | 1 |
| findExtrinsicCameraParams2 | 1 | 1 | 42 | 20 | 3.015 | 1.158 | 2 | 1 |
| BundleAdjusterBase::estimate | 19 | 1 | 2100 | 38 | 64.32 | 3.224 | 55 | 4 |
| findHomography | 1 | 1 | 21 | 10 | 7.553 | 2.267 | 5 | 1 |

</details>

### Other changes in this PR

* Stereo calibration was broken during 4.x to 5.x porting, fixing it
* Temporary fixes for `Submap` class and related stuff (anyway it'll be done using updated `PoseGraph` class)
* minor changes

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/3d/include/opencv2/3d.hpp [hunk 1]
```diff
@@ -468,77 +468,258 @@ can be found in:
...
+To do that, at each iteration it repeatedly calculates the energy at probe points until it's reduced.
+To calculate a probe point, a linear equation is solved: (J^T*J + lambda*D)*dx = -J^T*b where J is a function jacobian,
+b is a vector of residuals (aka errors or energy terms), D is a diagonal matrix generated from J^T*J diagonal
+and lambda changes for each probe point. Then the resulting dx is "added" to current variable and it forms
+a probe value. "Added" is quoted because in some groups (e.g. SO(3) group) such an increment can be a non-trivial operation.
+
+For more details, please refer to Wikipedia page (https://en.wikipedia.org/wiki/Levenberg%E2%80%93Marquardt_algorithm).
 */
-class CV_EXPORTS LMSolver : public Algorithm
+class CV_EXPORTS BaseLevMarq
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:489]: BaseLevMarq -> LevMarqBase or LevMarqSolver

File: modules/3d/include/opencv2/3d.hpp [hunk 2]
```diff
@@ -468,77 +468,258 @@ can be found in:
  */
 CV_EXPORTS_W void Rodrigues( InputArray src, OutputArray dst, OutputArray jacobian = noArray() );
 
-/** Levenberg-Marquardt solver. Starting with the specified vector of parameters it
-    optimizes the target vector criteria "err"
-    (finds local minima of each target vector component absolute value).
 
-    When needed, it calls user-provided callback.
+/** Base class for Levenberg-Marquadt solvers.
+This class can be used for general local optimization using sparse linear solvers, exponential param update or fixed variables
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:473]:
```
@brief one liner
<empty line>
details
```

File: modules/3d/include/opencv2/3d.hpp [hunk 3]
```diff
@@ -468,77 +468,258 @@ can be found in:
...
+    unsigned int maxIterations;
+
+    // LevMarq up and down params
+    double initialLambdaLevMarq;
+    double initialLmUpFactor;
+    double initialLmDownFactor;
+
+    Ptr<Backend> pBackend;
+
+    BaseLevMarq(Ptr<Backend> backend_) :
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:586]: Again, `const reference` for all input parameter.

---

no underscores in names if we don't need them.

File: modules/3d/include/opencv2/3d.hpp [hunk 4]
```diff
@@ -468,77 +468,258 @@ can be found in:
...
+    // optimization stops when norm2(dx) drops below this value
+    double stepNormTolerance;
+    // optimization stops when relative energy change drops below this value
+    double relEnergyDeltaTolerance;
+    // optimization stops when max gradient value (J^T*b vector) drops below this value
+    double minGradientTolerance;
+    // optimization stops when energy drops below this value
+    double smallEnergyTolerance;
+    // optimization stops after a number of iterations performed
+    unsigned int maxIterations;
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:577]: No fields in public API.
This is the primary requirement of public OpenCV API.
getter/setter should be exposed instead.

File: modules/3d/include/opencv2/3d.hpp [hunk 5]
```diff
@@ -468,77 +468,258 @@ can be found in:
...
+    {
+        Report(bool found_, int iters_, double energy_) :
+            found(found_), iters(iters_), energy(energy_)
+        { }
+        bool found;
+        int iters;
+        double energy;
+    };
+
+    class Backend
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:502]: Looks like it is not used by user code directly. Move to src or details.

File: modules/3d/include/opencv2/3d.hpp [hunk 6]
```diff
@@ -468,77 +468,258 @@ can be found in:
...
 public:
-    class CV_EXPORTS Callback
+    struct Report
+    {
+        Report(bool found_, int iters_, double energy_) :
+            found(found_), iters(iters_), energy(energy_)
+        { }
+        bool found;
+        int iters;
+        double energy;
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:499]: missing documentation

File: modules/core/include/opencv2/core/dualquaternion.inl.hpp [hunk 7]
```diff
@@ -158,7 +158,7 @@ inline Quat<T> DualQuat<T>::getRotation(QuatAssumeType assumeUnit) const
 template <typename T>
 inline Vec<T, 3> DualQuat<T>::getTranslation(QuatAssumeType assumeUnit) const
 {
-    Quat<T> trans = 2.0 * (getDualPart() * getRealPart().inv(assumeUnit));
+    Quat<T> trans = T(2.0) * (getDualPart() * getRealPart().inv(assumeUnit));
```
Comment (alalek) [inline modules/core/include/opencv2/core/dualquaternion.inl.hpp:161]: Such changes should be backported to 4.x

File: modules/calib/src/calibration.cpp [hunk 8]
```diff
@@ -1045,29 +1052,44 @@ static double stereoCalibrateImpl(
...
+    };
+
+    double reprojErr = 0;
+    if (countNonZero(mask))
+    {
+        LevMarqDenseLinear solver(param, lmcallback, mask);
+        // old LMSolver calculates successful iterations only, this one calculates all iterations
+        solver.maxIterations = (unsigned int)(termCrit.maxCount * 2.1);
+        solver.stepNormTolerance = termCrit.epsilon;
+        solver.smallEnergyTolerance = termCrit.epsilon * termCrit.epsilon;
```
Comment (alalek) [inline modules/calib/src/calibration.cpp:1069]: We should have setter with TermCriteria parameter.

> maxIterations = (unsigned int)(termCrit.maxCount * 2.1);
> 2.1

Why?

**Review by alalek — COMMENTED** [review]
We should revise public API.
Keep minimal part which is necessary for user application (tests).

Internal interfaces like "BaseLevMarq::Backend" should be hidden.

File: modules/calib/src/calibration.cpp (hunk 8 above)
Comment (savuor) [reply modules/calib/src/calibration.cpp:1069]: `2.1` is a compatibility hack. Old impl counts successful iterations only, the new one counts all iterations.
The proportion between them for the same run is different but in average it is 2.1: `newIters = oldIters*2.1`

I propose three solutions here:
* to leave this hack as is (adding a comment about it in src)
* to find true multiplier for each use case based on iterations elapsed statistics
* remove the multiplier: it's 5.x now, we don't have to maintain such parameters compatibility

File: modules/3d/include/opencv2/3d.hpp (hunk 5 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp:502]: The motivation to leave the `Backend` in public was to let a user to construct solvers that use sparse matrices, params of SE(3) and of more exotic groups, with fixed variables and so on.
I will try to move it to details to keep this possibility.
(However the interface is very complicated for a user)

File: modules/calib/src/calibration.cpp (hunk 8 above)
Comment (vpisarev) [reply modules/calib/src/calibration.cpp:1069]: I think, having such a hack is fine.
1. If the problem solved is more or less well-defined (not ill-posed), the solver should converge faster than it will reach the maximum. 
2. If the problem solved is ill-posed then we will do at least termCrit.maxCount iterations, so this hack may affect the speed, but not the quality. But, I believe, because the new solver uses the sparse structure of matrices, it should run even faster than the previous "dense" implementation.

File: modules/3d/include/opencv2/3d.hpp (hunk 4 above)
Comment (vpisarev) [reply modules/3d/include/opencv2/3d.hpp:577]: can we try to utilize https://github.com/opencv/opencv/wiki/OE-34.-Named-Parameters. It lets to define very nice API for C++ and Python (and maybe Swift after the corresponding binding generator is revised)

File: modules/calib/src/calibration.cpp (hunk 8 above)
Comment (savuor) [reply modules/calib/src/calibration.cpp:1069]: Multiplier removed, statistics recalculated.
Regarding `TermCriteria` usage: I'm against it, it has only one epsilon parameter and it's not obvious to what threshold in this LevMarq impl it corresponds to.

File: modules/3d/include/opencv2/3d.hpp (hunk 1 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: Done

File: modules/3d/include/opencv2/3d.hpp (hunk 2 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp:474]: Fixed

File: modules/3d/include/opencv2/3d.hpp (hunk 3 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: fixed

File: modules/3d/include/opencv2/3d.hpp (hunk 4 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: Implemented it as a `Setting` structure

File: modules/3d/include/opencv2/3d.hpp (hunk 5 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: `Backend` is moved to details header

File: modules/3d/include/opencv2/3d.hpp (hunk 6 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: Added docs for `Report` structure

File: modules/core/include/opencv2/core/dualquaternion.inl.hpp (hunk 7 above)
Comment (savuor) [reply modules/core/include/opencv2/core/dualquaternion.inl.hpp:161]: Backported to #21319 and [#3137@contrib](https://github.com/opencv/opencv_contrib/pull/3137)

File: modules/3d/include/opencv2/3d.hpp [hunk 9]
```diff
@@ -468,77 +468,308 @@ can be found in:
...
+            relEnergyDeltaTolerance(1e-6),
+            minGradientTolerance(1e-6),
+            smallEnergyTolerance(0), // not used by default
+            maxIterations(500),
+            initialLambda(0.0001),
+            initialLmUpFactor(2.0),
+            initialLmDownFactor(3.0)
+        { }
+
+        bool operator==(const Settings& other) const
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:545]: If users don't need this functionality, then it is better to create internal function.

File: modules/3d/include/opencv2/3d.hpp [hunk 10]
```diff
@@ -468,77 +468,308 @@ can be found in:
...
+    MySolver solver(nVars, callback, MySolver::Settings().geodesicS(true).geoScale(1.0));
+    @endcode
+
+    */
+    struct CV_EXPORTS Settings
+    {
+        Settings() :
+            jacobiScaling(false),
+            upDouble(true),
+            useStepQuality(true),
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:526]: Move ctor implementation to .cpp file.

File: modules/3d/include/opencv2/3d.hpp [hunk 11]
```diff
@@ -468,77 +468,308 @@ can be found in:
...
+            ok = ok && abs(this->initialLambda - other.initialLambda) < eps;
+            ok = ok && abs(this->initialLmUpFactor - other.initialLmUpFactor) < eps;
+            ok = ok && abs(this->initialLmDownFactor - other.initialLmDownFactor) < eps;
+
+            return ok;
+        }
+
+        Settings& jacobiScalingS          (bool         v) { jacobiScaling           = v; return *this; }
+        Settings& upDoubleS               (bool         v) { upDouble                = v; return *this; }
+        Settings& useStepQualityS         (bool         v) { useStepQuality          = v; return *this; }
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:578]: Why is here 'S' suffix?
Where are you find that?

File: modules/3d/include/opencv2/3d.hpp [hunk 12]
```diff
@@ -468,77 +468,308 @@ can be found in:
...
+            ok = ok && (this->maxIterations == other.maxIterations);
+
+            ok = ok && abs(this->initialLambda - other.initialLambda) < eps;
+            ok = ok && abs(this->initialLmUpFactor - other.initialLmUpFactor) < eps;
+            ok = ok && abs(this->initialLmDownFactor - other.initialLmDownFactor) < eps;
+
+            return ok;
+        }
+
+        Settings& jacobiScalingS          (bool         v) { jacobiScaling           = v; return *this; }
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:576]: `inline` for all setters

File: modules/3d/include/opencv2/3d.hpp [hunk 13]
```diff
@@ -468,77 +468,308 @@ can be found in:
...
+        Settings& checkMinGradientS       (bool         v) { checkMinGradient        = v; return *this; }
+        Settings& checkStepNormS          (bool         v) { checkStepNorm           = v; return *this; }
+        Settings& geodesicS               (bool         v) { geodesic                = v; return *this; }
+        Settings& hGeoS                   (double       v) { hGeo                    = v; return *this; }
+        Settings& geoScaleS               (double       v) { geoScale                = v; return *this; }
+        Settings& stepNormToleranceS      (double       v) { stepNormTolerance       = v; return *this; }
+        Settings& relEnergyDeltaToleranceS(double       v) { relEnergyDeltaTolerance = v; return *this; }
+        Settings& minGradientToleranceS   (double       v) { minGradientTolerance    = v; return *this; }
+        Settings& smallEnergyToleranceS   (double       v) { smallEnergyTolerance    = v; return *this; }
+        Settings& maxIterationsS          (unsigned int v) { maxIterations           = v; return *this; }
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:591]: > unsigned int

This type may be non-friendly with bindings.

File: modules/3d/include/opencv2/3d.hpp [hunk 14]
```diff
@@ -468,77 +468,308 @@ can be found in:
...
-         @param err output vector of errors: err_i = actual_f_i - ideal_f_i
-         @param J output Jacobian: J_ij = d(err_i)/d(param_j)
-
-         when J=noArray(), it means that it does not need to be computed.
-         Dimensionality of error vector and param vector can be different.
-         The callback should explicitly allocate (with "create" method) each output array
-         (unless it's noArray()).
-        */
-        virtual bool compute(InputArray param, OutputArray err, OutputArray J) const = 0;
+        virtual ~Backend() { }
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:642]: > Defined in

Is it not a definition?

Why does forward declaration not work here?

File: modules/3d/include/opencv2/3d.hpp [hunk 15]
```diff
@@ -468,77 +468,308 @@ can be found in:
...
     };
 
-    /**
-       Runs Levenberg-Marquardt algorithm using the passed vector of parameters as the start point.
-       The final vector of parameters (whether the algorithm converged or not) is stored at the same
-       vector. The method returns the number of iterations used. If it's equal to the previously specified
-       maxIters, there is a big chance the algorithm did not converge.
+    LevMarqBase(const Ptr<Backend>& backend, const Settings& settings);
```
Comment (alalek) [inline modules/3d/include/opencv2/3d.hpp:645]: Users don't really need this constructor. As they can't use it.

File: modules/3d/include/opencv2/3d.hpp (hunk 11 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp:578]: S stands for "set" but I didn't want to give them names like `setEpsilon` to make method names more compact.
Since you recommended not to use underscores, I seek other ways to name them.

File: modules/3d/include/opencv2/3d.hpp (hunk 9 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: Removed
(was made to compare a passed arg with default settings constant, default arg is used instead)

File: modules/3d/include/opencv2/3d.hpp (hunk 10 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: Done

File: modules/3d/include/opencv2/3d.hpp (hunk 11 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: Decided to use `setValue` names instead, looks more natural

File: modules/3d/include/opencv2/3d.hpp (hunk 12 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: Done

File: modules/3d/include/opencv2/3d.hpp (hunk 13 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: Replaced by int

File: modules/3d/include/opencv2/3d.hpp (hunk 14 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: `Backend` class removed

File: modules/3d/include/opencv2/3d.hpp (hunk 15 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp]: Fixed, see the comment below

**Comment by savuor** [comment]
1. `LevMarqBase` was moved to `detail` headers
2. `LevMarqDenseLinear` was replaced by a class `LevMarq` which takes enum args responsible for "dense" and "linear" properties respectively; other types of solvers are not implemented now and reserved for future.

**Review by alalek — APPROVED** [review]
Looks good to me :+1:
